flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotAllocator.java

```diff
@@ -54,6 +54,18 @@ public interface SlotAllocator {
     Optional<? extends VertexParallelism> determineParallelism(
             JobInformation jobInformation, Collection<? extends SlotInfo> slots);
 
+    /**
+     * Same as {@link #determineParallelism(JobInformation, Collection)} but additionally determine
+     * assignment of slots to execution slot sharing groups.
+     */
+    default Optional<? extends VertexParallelismWithSlotSharing>
```

Comment:
> ? extends VertexParallelismWithSlotSharing

The existence and usage of the `VertexParallelismWithSlotSharing` is an implementation detail of the `SlotSharingSlotAllocation`. Ideally this wouldn't be exposed.

Essentially you are exposing the concept of slot sharing to other components, and force all slot allocator implementations to be aware of it.

Reply:
I agree that slot sharing is an implementation detail.
However, one difference that I'd like to expose, is that `VertexParallelismWithSlotSharing` contains assignments, while `VertexParallelism` is only about parallelism.

I think the right way to do that is to parameterize `SlotAllocator` with this type, and let `SlotSharingSlotAllocator` use `VertexParallelismWithSlotSharing`:
```
interface SlotAllocator<A> {
  VertexParallelism determineParallelism(); // for hasResources, canScaleUp
  A determineAssignments(...); // for tryReserveResources later
  tryReserveResources(A);
}
```
WDYT?

Reply:
After an offline discussion, I restructured the interfaces (4fa93e3756fad77cf2dbc92bf77614aa39cc28fe..2fc84d83884c1f342d46fec02d961d4f73eb4219).

This particular comment is addressed by 4fa93e3756fad77cf2dbc92bf77614aa39cc28fe:
```
Restructure 1/5: change types passed between AdaptiveScheduler and SlotAssigner

Slot assignments are computed and consumed by SlotAllocator.
This is expressed implicitly by implementing VertexParallelism.

This change tries to make that clear, while still allowing to assign
slots to something other than slot sharing groups.

It does so by:
1. Introduce JobSchedulingPlan, computed and consumed by SlotAllocator. It couples VertexParallelism with slot assignments.
2. Push the polymorphism of state assignments from VertexParallelism to slot assignment target in JobSchedulingPlan.
```